rewrite-core/src/main/java/org/openrewrite/shell/exec/ShellExecutor.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2024 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.shell.exec;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Incubating;
+import org.openrewrite.scheduling.WorkingDirectoryExecutionContextView;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+@Incubating(since = "8.30.0")
+@SuppressWarnings("unused")
+public interface ShellExecutor {
+
+    @SuppressWarnings("unused")
+    default void init() {}
+
+    @SuppressWarnings("unused")
+    default Path exec(List<String> command, Path workingDirectory, Map<String, String> environment, ExecutionContext ctx) {
+        return exec(command, workingDirectory, environment, ctx, 5L);
+    }
+
+    default Path exec(List<String> command, Path workingDirectory, Map<String, String> environment, ExecutionContext ctx, Long timeoutMinutes) {
```

Comment:
I'd prefer a Duration for the timeout here, as opposed to a Long; and elsewhere we're fairly consistent in having `ExecutionContext` be the last argument; let's do that here as well.

Reply:
Will do. Thanks @timtebeek!